main.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/bitrise-io/go-steputils/cache"
+	"github.com/bitrise-io/go-steputils/stepconf"
+	"github.com/bitrise-io/go-utils/command"
+	"github.com/bitrise-io/go-utils/errorutil"
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/kballard/go-shellquote"
+)
+
+type config struct {
+	WorkingDir  string `env:"workdir,dir"`
+	YarnCommand string `env:"command"`
+	YarnArgs    string `env:"args"`
+	UseCache    string `env:"cache_local_deps,opt[yes,no]"`
+	IsDebugLog  string `env:"verbose_log,opt[yes,no]"`
+}
+
+func failf(format string, v ...interface{}) {
+	log.Errorf(format, v...)
+	os.Exit(1)
+}
+
+func getInstallYarnCommand() (*command.Model, error) {
+	if runtime.GOOS != "linux" {
+		return nil, fmt.Errorf("Unsupported platform %s, failed to install yarn", runtime.GOOS)
+	}
+	if _, err := os.Stat(path.Join("etc", "lsb-release")); err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("only Ubuntu distribution supported, failed to install yarn")
+		}
+		return nil, err
+	}
+
+	installCmd := command.New("sh", "-c", `curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
+echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
+sudo apt-get update && sudo apt-get install -y yarn`)
+	installCmd.SetStdout(os.Stdout).SetStderr(os.Stderr)
+
+	return installCmd, nil
+}
+
+func cacheYarn(workingDir string) error {
+	yarnCache := cache.New()
+	var cachePaths []string
+
+	// Supporting yarn workspaces (https://yarnpkg.com/lang/en/docs/workspaces/), for this recursively look
+	// up all node_modules directories
+	if err := filepath.Walk(workingDir, func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if fileInfo.IsDir() && fileInfo.Name() == "node_modules" {
+			cachePaths = append(cachePaths, path)
+			return filepath.SkipDir
+		}
+		return nil
+	}); err != nil {
+		failf("Failed to find node_modules directories, error: %s", err)
```

Comment:
- You should not use `failf` in a method, just return the error and use `failf` in the `main()`.
- I think the step should not fail if the cache failed. I would write a warning only. WDYT? 🤔

Reply:
Agree with the warning, changed the code.